File: hashing/counting/c++/permutation_string/permutation_string.cpp
// permutation_string.cpp
#include "permutation_string.hpp"
#include <string>
#include <unordered_map>

bool checkInclusion(const std::string &s1, const std::string &s2)
{
    int s1Len = s1.size(), s2Len = s2.size();
    if (s1Len > s2Len)
    {
        return false;
    }
    std::unordered_map<char, int> s1Counts;
    std::unordered_map<char, int> s2WindowCounts;
    // O(n) where n is the length of s1
    for (int i = 0; i < s1Len; i++)
    {
        auto s1Result = s1Counts.try_emplace(s1[i], 1);
        auto s2Result = s2WindowCounts.try_emplace(s2[i], 1);
        if (!s1Result.second)
        {
            s1Result.first->second++;
        }
        if (!s2Result.second)
        {
            s2Result.first->second++;
        }
    }

    // O(m) where m = s2.size() - s1.size()
    for (int windowEnd = s1Len; windowEnd < s2Len; windowEnd++)
    {
        // O(n) average
        if (s1Counts == s2WindowCounts)
        {
            return true;
        }
        // Add the next character or increment its count if already exists
        s2WindowCounts[s2[windowEnd]]++;
        // To keep the window fixed to len(s1), decrement the count of the element at (i - len(s1))
        int windowStart = windowEnd - s1Len;
        s2WindowCounts[s2[windowStart]]--;
        // If the count of `s2[windowStart]` reaches zero, delete the key so the comparison is more efficient
        if (s2WindowCounts[s2[windowStart]] == 0)
        {
            s2WindowCounts.erase(s2[windowStart]);
        }
    }
    // Last comparison is needed since the == check happens before inserting, incrementing, decrementing counts
    // The last iteration of the loop finished without comparing the last window of length len(s1) in s2
    return s1Counts == s2WindowCounts;
}

File: hashing/counting/python3/permutation_string.py
from collections import Counter


def check_inclusion(s1: str, s2: str) -> bool:
    """
    Given two strings `s1` and `s2`, return `True` if `s2` contains a permutation of `s1`.

    Parameters
    ----------
    s1 : str
        A string consisting of lower case english characters
    s2 : str
        A string consisting of lower case english characters

    Returns
    -------
    bool
        `True` if one of `s1`'s permutation is a substring of `s2`, `False` otherwise
    """
    len_s1 = len(s1)
    len_s2 = len(s2)
    if len_s1 > len_s2:
        return False
    # O(n) to build where n = len(s1)
    s1_counts = Counter(s1)
    # Initialized with first len(s1) characters of s2
    s2_window_counts = Counter(s2[slice(len_s1)])
    # O(m) where m = len(2) - len(s1)
    for window_end in range(len_s1, len_s2):
        # O(n) comparisons if every character in s1 is unique and both hashmaps contain n keys
        if s1_counts == s2_window_counts:
            return True
        # Add the next character or increment its count if already exists
        s2_window_counts[s2[window_end]] += 1
        # To keep the window fixed to len(s1), decrement the count of the element at (pointer - len(s1))
        window_start = window_end - len_s1
        s2_window_counts[s2[window_start]] -= 1
        # If the count of `s2[window_start]` reaches zero, delete the key so the comparison is more efficient
        if s2_window_counts[s2[window_start]] == 0:
            del s2_window_counts[s2[window_start]]
    # Last comparison is needed since the == check happens before inserting, incrementing, decrementing counts
    # The last iteration of the loop finished without comparing the last window of length len(s1) in s2
    return s1_counts == s2_window_counts


def main() -> int:
    for s1, s2 in [
        ["a", "ab"],
        ("ab", "eidbaooo"),
        ["ab", "eidboaoo"],
        ("adc", "dcda"),
    ]:
        contains_perm = check_inclusion(s1, s2)
        print(
            f"The string '{s2}' {'contains' if contains_perm else 'does not contain'} a permutation of the string '{s1}'"
        )

    return 0


if __name__ == "__main__":
    main()
